pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -0,0 +1,344 @@
+		changed = true
+	}
+
+	if changed {
+		if _, err = e.client.ServiceAccounts(liveServiceAccount.Namespace).Update(liveServiceAccount); err != nil {
+			return false, err
+		}
+	}
+	return false, nil
+}
+
+const (
+	tokenSecretWaitInterval = 100 * time.Millisecond
+	tokenSecretWaitTimes    = 20
+)
+
+// createTokenSecret creates a token secret for a given service account.  Returns the name of the token
+func (e *DockercfgController) createTokenSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
+	tokenSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      secret.Strategy.GenerateName(getTokenSecretNamePrefix(serviceAccount)),
+			Namespace: serviceAccount.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey: serviceAccount.Name,
+				api.ServiceAccountUIDKey:  string(serviceAccount.UID),
+			},
+		},
+		Type: api.SecretTypeServiceAccountToken,
+		Data: map[string][]byte{},
+	}
+
+	_, err := e.client.Secrets(tokenSecret.Namespace).Create(tokenSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	// now we have to wait for the service account token controller to make this valid
+	// TODO remove this once we have a create-token endpoint
+	for i := 0; i <= tokenSecretWaitTimes; i++ {
+		liveTokenSecret, err := e.client.Secrets(tokenSecret.Namespace).Get(tokenSecret.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if _, exists := liveTokenSecret.Data[api.ServiceAccountTokenKey]; exists {
+			return liveTokenSecret, nil
+		}
+
+		time.Sleep(wait.Jitter(tokenSecretWaitInterval, 0.0))
+
+	}
+
+	// the token wasn't ever created, attempt deletion
+	if err := e.client.Secrets(tokenSecret.Namespace).Delete(tokenSecret.Name); err != nil {
+		glog.Error(err)
+	}
+	return nil, fmt.Errorf("token never generated for %s", tokenSecret.Name)
+}
+
+// createDockerPullSecret creates a dockercfg secret based on the token secret
+func (e *DockercfgController) createDockerPullSecret(serviceAccount *api.ServiceAccount) (*api.Secret, error) {
+	tokenSecret, err := e.createTokenSecret(serviceAccount)
+	if err != nil {
+		return nil, err
+	}
+
+	dockercfgSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      secret.Strategy.GenerateName(getDockercfgSecretNamePrefix(serviceAccount)),
+			Namespace: tokenSecret.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey:        serviceAccount.Name,
+				api.ServiceAccountUIDKey:         string(serviceAccount.UID),
+				ServiceAccountTokenSecretNameKey: string(tokenSecret.Name),
+			},
+		},
+		Type: api.SecretTypeDockercfg,
+		Data: map[string][]byte{},
+	}
+
+	// prevent updating the DockerURL until we've created the secret
+	e.dockerURLLock.Lock()
+	defer e.dockerURLLock.Unlock()
+
+	dockercfg := &credentialprovider.DockerConfig{
+		e.dockerURL: credentialprovider.DockerConfigEntry{
+			Username: "serviceaccount",
```

Comment:
@ncdc just making sure nothing in our registry is depending on the username or email any more